platform-operator/controllers/verrazzano/reconcile/rancher.go

```
@@ -0,0 +1,139 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// createRancherIngressAndCertCopies - creates copies of Rancher ingress and cert secret if
+// they exist. If there is a DNS update in progress, we want to do this so that the old DNS
+// continues to work as long as there is a good DNS entry, so that managed clusters can continue
+// connecting using the old Rancher DNS in their kubeconfig, until they are updated with the new DNS.
+func (r *Reconciler) createRancherIngressAndCertCopies(ctx spi.ComponentContext) {
+	ok, rancherComp := registry.FindComponent(rancher.ComponentName)
+	if !ok {
+		return
+	}
+
+	dnsSuffix := getDNSSuffix(ctx.EffectiveCR())
+	if dnsSuffix == "" {
+		ctx.Log().Debug("Empty DNS suffix, skipping Rancher ingress copy")
+		return
+	}
+
+	for _, ingressName := range rancherComp.GetIngressNames(ctx) {
+		ing := netv1.Ingress{}
+		err := r.Get(context.TODO(), ingressName, &ing)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				ctx.Log().Infof("Cannot make a copy of Rancher ingress %s/%s - get ingress failed: %v", ingressName.Namespace, ingressName.Name, err)
+			}
+			continue
+		}
+
+		if len(ing.Spec.TLS) == 0 || len(ing.Spec.TLS[0].Hosts) == 0 {
+			continue
+		}
+		if strings.HasSuffix(ing.Spec.TLS[0].Hosts[0], dnsSuffix) {
+			ctx.Log().Debugf("Rancher ingress host %s has DNS suffix %s, skipping copy", ing.Spec.TLS[0].Hosts[0], dnsSuffix)
+			continue
+		}
+
+		newIngressTLSList := r.createIngressCertSecretCopies(ctx, ing)
+		newIngressNSN := types.NamespacedName{Name: "vz-" + ing.Name, Namespace: ingressName.Namespace}
+		err = r.createIngressCopy(newIngressNSN, ing, newIngressTLSList)
+		if err != nil {
+			ctx.Log().Infof("Failed to create a copy of Rancher ingress %s/%s - create ingress failed: %v", ingressName.Namespace, ingressName.Name, err)
+			continue
```

What is the justification for ignoring this error?

---

And the ones prior

---

I think the idea I had was that this was not something that should fail the Rancher install/update and hence the VZ install/update. This copying is needed only in a subset of cases and there is a workaround i.e. manually reapplying the registration.
Of course, the downside is that diagnosability when multicluster does not work is made a bit more difficult. I am open to opinions on whether we should make this a failing change.

---

I had thought about this too, we'd have to make `beforeInstallComponents` return an error and then requeue and not proceed with installing components until it passes. I've gone back and forth, but I think I'm leaning towards making this a "best effort" and not fail the install. We can always make it more restrictive in the future.